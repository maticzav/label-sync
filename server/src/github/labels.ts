/**
 * This file contains utility functions that we use to communicate
 * with GitHub about issues and related content.
 */

import { ProbotOctokit } from 'probot'

import { withDefault } from '../utils'

type Octokit = InstanceType<typeof ProbotOctokit>

// MARK: - Types

export interface GithubLabel {
  /* Naming */
  old_name?: string
  name: string
  /* Description */
  old_description?: string | null
  description?: string | null
  /* Colors */
  old_color?: string
  color: string
  default?: boolean
}

// MARK: - Methods

type GetRepositoryLabelsParams = {
  repo: string
  owner: string
}

/**
 * Fetches labels in a repository.
 */
export async function getRepositoryLabels(
  github: Octokit,
  params: GetRepositoryLabelsParams,
) {
  let labels: GithubLabel[] = []

  /* Github paginates from page 1 */
  await handler(1)

  return labels

  /* Paginates and performs changes. */
  async function handler(page: number) {
    const repoLabels = await github.issues
      .listLabelsForRepo({
        owner: params.owner,
        repo: params.repo,
        per_page: 100,
        page: page,
      })
      .then((res) => res.data)

    labels.push(...repoLabels)

    /* Rerun handler if there are more labels available. */
    /* istanbul ignore next */
    if (repoLabels.length === 100) {
      await handler(page + 1)
    }
  }
}

type AddLabelsToRepositoryParams = {
  repo: string
  owner: string
  /**
   * List of labels that we should add to the repository.
   */
  labels: Pick<GithubLabel, 'name' | 'color'>[]
  /**
   * Tells whether this is only a simulation or persisting change.
   */
  persist: boolean
}

/**
 * Create new labels in a repository.
 */
export async function addLabelsToRepository(
  github: Octokit,
  params: AddLabelsToRepositoryParams,
): Promise<GithubLabel[]> {
  /* Return immediately on non-persistent sync. */
  if (params.persist === false) return params.labels

  /* Perform sync on persist. */
  const actions = params.labels.map((label) => addLabelToRepository(label))
  await Promise.all(actions)

  return params.labels

  /**
   * Helper functions
   */
  async function addLabelToRepository(label: GithubLabel) {
    try {
      await github.issues.createLabel({
        owner: params.owner,
        repo: params.repo,
        name: label.name,
        description: label.description || undefined,
        color: label.color,
      })
    } catch (err) /* istanbul ignore next */ {
      throw new Error(
        `Couldn't create ${label.name} in ${params.owner}/${params.repo}: ${err.message}`,
      )
    }
  }
}

type UpdateLabelsInRepositoryParams = {
  repo: string
  owner: string
  /**
   * Labels that we should update.
   */
  labels: GithubLabel[]
  /**
   * Tells whether the changes should remain visible.
   */
  persist: boolean
}

/**
 *
 * Updates labels in repository.
 *
 * When old_name is specified in the label, we try to rename the label.
 */
export async function updateLabelsInRepository(
  github: Octokit,
  params: UpdateLabelsInRepositoryParams,
): Promise<GithubLabel[]> {
  /* Return immediately on non-persistent sync. */
  if (params.persist === false) return params.labels

  /* Update values on persist. */
  const actions = params.labels.map((label) => updateLabelInRepository(label))
  await Promise.all(actions)

  return params.labels

  /**
   * Helper functions
   */
  async function updateLabelInRepository(label: GithubLabel) {
    try {
      await github.issues.updateLabel({
        current_name: withDefault(label.name, label.old_name),
        owner: params.owner,
        repo: params.repo,
        name: label.name,
        description: label.description || undefined,
        color: label.color,
      })
    } catch (err) /* istanbul ignore next */ {
      throw new Error(
        `Couldn't update ${label.name} in ${params.owner}/${params.repo}: ${err.message}`,
      )
    }
  }
}

type RemoveLabelFromRepositoryParams = {
  repo: string
  owner: string
  /**
   * Labels that we should remove.
   */
  labels: Pick<GithubLabel, 'name'>[]
  /**
   * Tells whether the changes should be made live.
   */
  persist: boolean
}

/**
 * Removes labels from repository.
 */
export async function removeLabelsFromRepository(
  github: Octokit,
  params: RemoveLabelFromRepositoryParams,
): Promise<Pick<GithubLabel, 'name'>[]> {
  /* Return immediately on non-persistent sync. */
  if (params.persist === false) return params.labels

  const actions = params.labels.map((label) =>
    removeLabelFromRepository(label.name),
  )
  await Promise.all(actions)

  return params.labels

  /**
   * Helper functions
   */
  async function removeLabelFromRepository(label: string) {
    try {
      const ghLabel = await github.issues.deleteLabel({
        owner: params.owner,
        repo: params.repo,
        name: label,
      })

      return ghLabel.data
    } catch (err) /* istanbul ignore next */ {
      throw new Error(
        `Couldn't remove ${label} in ${params.owner}/${params.repo}: ${err.message}`,
      )
    }
  }
}

type AliasLabelsInRepositoryParams = {
  repo: string
  owner: string
  /**
   * Labels to alias.
   */
  labels: GithubLabel[]
  /**
   * Tells whether we should apple the changes or just simulate them.
   */
  persist: boolean
}

/**
 * Aliases labels in a repository by adding them to labels.
 */
export async function aliasLabelsInRepository(
  github: Octokit,
  params: AliasLabelsInRepositoryParams,
): Promise<GithubLabel[]> {
  let page = 1

  /* Skip on no labels */
  if (params.labels.length === 0) return []

  await handler()

  return params.labels

  /* Paginates and performs changes. */
  async function handler() {
    const issues = await github.issues
      .listForRepo({
        owner: params.owner,
        repo: params.repo,
        per_page: 100,
        page,
      })
      .then((res) => res.data)

    /* Process all the issues. */
    for (const issue of issues) {
      /* Filter labels that should be in this issue but are not. */
      const missingLabels = params.labels.filter((label) =>
        issue.labels.some((issueLabel) => issueLabel.name === label.old_name),
      )

      if (missingLabels.length === 0) continue

      /* Add all the missing labels. */
      await addLabelsToIssue(github, {
        ...params,
        issue: issue.number,
        labels: missingLabels,
      })
    }

    /* Rerun handler if there are more issues available. */
    /* istanbul ignore next */
    if (issues.length === 100) {
      page += 1
      await handler()
    }
  }
}

type AddLabelsToIssueParams = {
  repo: string
  owner: string
  issue: number
  /**
   * Names of the labels to add to the issue.
   */
  labels: Pick<GithubLabel, 'name'>[]
  /**
   * Tells whether we should only simulate the function.
   */
  persist: boolean
}

/**
 * Adds labels to an issue.
 */
export async function addLabelsToIssue(
  github: Octokit,
  params: AddLabelsToIssueParams,
): Promise<Pick<GithubLabel, 'name'>[]> {
  /* Return immediately on non-persistent sync. */
  /* istanbul ignore next */
  if (params.persist === false) return params.labels

  const ghLabels = await github.issues
    .addLabels({
      repo: params.repo,
      owner: params.owner,
      issue_number: params.issue,
      labels: params.labels.map((label) => label.name),
    })
    .then((res) => res.data)

  return ghLabels
}

/**
 * Compares two labels by comparing all of their keys.
 */
export function isLabel(local: GithubLabel): (remote: GithubLabel) => boolean {
  return (remote) =>
    local.name === remote.name &&
    local.description === remote.description &&
    local.color === remote.color
}

/**
 * Determines whether the two configurations configure the same label.
 */
export function isLabelDefinition(
  local: GithubLabel,
): (remote: GithubLabel) => boolean {
  return (remote) => local.name === remote.name
}
