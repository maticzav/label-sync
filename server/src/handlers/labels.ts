import { Octokit } from 'probot'

import {
  LSCRepository,
  LSCLabel,
  LSCRepositoryConfiguration,
} from '../configuration'
import * as maybe from '../data/maybe'
import { Dict } from '../data/dict'
import {
  GithubLabel,
  getRepositoryLabels,
  addLabelsToRepository,
  updateLabelsInRepository,
  removeLabelsFromRepository,
  aliasLabelsInRepository,
} from '../github'
import { withDefault } from '../utils'

export type LabelSyncReport =
  | {
      status: 'Success'
      owner: string
      repo: string
      additions: GithubLabel[]
      updates: GithubLabel[]
      removals: GithubLabel[]
      aliases: GithubLabel[]
      config: {
        config: Required<LSCRepositoryConfiguration>
        labels: Dict<LSCLabel>
      }
    }
  | {
      status: 'Failure'
      repo: string
      owner: string
      message: string
      config: {
        config: LSCRepositoryConfiguration
        labels: Dict<LSCLabel>
      }
    }

/**
 *
 * Performs a labels sync with configuration.
 *
 * @param client
 * @param repository
 * @param config
 * @param options
 */
export async function handleLabelSync(
  octokit: Octokit,
  owner: string,
  repo: string,
  { config, labels }: LSCRepository,
  persist: boolean,
): Promise<LabelSyncReport> {
  /**
   * Label Sync handler firstly loads current labels from Github,
   * and new labels from local configuration.
   *
   * After that it generates a diff and creates, updates or deletes
   * label definitions in a particular repository.
   */
  const remoteLabels = await getRepositoryLabels(octokit, { repo, owner })
  const labelsDiff = maybe.andThen(remoteLabels, calculateDiff(labels))

  /* Configuration */

  const removeUnconfiguredLabels = withDefault(
    false,
    config?.removeUnconfiguredLabels,
  )

  /* istanbul ignore if */
  if (labelsDiff === null) {
    return {
      status: 'Failure',
      owner,
      repo,
      message: `Couldn't make a diff of labels.`,
      config: {
        labels,
        config: {
          removeUnconfiguredLabels,
        },
      },
    }
  }

  const { added, changed, aliased, removed } = labelsDiff

  /* Perform sync */
  /**
   * 1. Add new labels.
   * 2. Perform updates.
   * 3. Alias labels.
   * 4. Remove labels.
   */

  const additions = await addLabelsToRepository(
    octokit,
    { repo, owner },
    added,
    persist,
  )
  const updates = await updateLabelsInRepository(
    octokit,
    { repo, owner },
    changed,
    persist,
  )
  const aliases = await aliasLabelsInRepository(
    octokit,
    { repo, owner },
    aliased,
    persist,
  )
  const removals = await removeLabelsFromRepository(
    octokit,
    { repo, owner },
    removed,
    removeUnconfiguredLabels && persist,
  )

  return {
    status: 'Success',
    repo,
    owner,
    additions,
    updates,
    removals,
    aliases,
    config: {
      config: {
        removeUnconfiguredLabels,
      },
      labels,
    },
  }
}

/**
 *
 * Calculates the diff of labels.
 *
 * @param currentLabels
 * @param newLabels
 */
export function calculateDiff(
  config: LSCRepository['labels'],
): (
  currentLabels: GithubLabel[],
) => {
  added: GithubLabel[]
  changed: GithubLabel[]
  aliased: GithubLabel[]
  removed: GithubLabel[]
} {
  /**
   * Label may fall into four classes:
   *  - it may be new: not on Github and no other label
   *    in the configuration references it
   *  - it may be updated: name remains the same but color
   *    or description have changed
   *  - it may be aliased: one of the labels references a
   *    label that exists on Github but is not present in
   *    the configuration anymore
   *  - it may be removed: no labels alias it and it is
   *    not in the configuration
   *
   * In light of that we diff labels so that:
   *  - added: contains all the labels we should add
   *  - changed: contains all the labels that we should rename,
   *      change description or change color
   *  - aliased: contains all the labels that we should cleverly
   *      reference in the issues
   *  - removed: contains all the labels that we should remove
   */

  return (currentLabels) => {
    const currentLabelsNames = currentLabels.map((l) => l.name)

    /* Labels */

    let added: GithubLabel[] = []
    let changed: GithubLabel[] = []
    let aliased: GithubLabel[] = []
    // let aliased: { [target: string]: GithubLabel[] } = {}
    let removed: GithubLabel[] = []

    /* Calculate differences */

    // TODO: you may not split label alias

    for (const label of Object.keys(config)) {
      /**
       * We know that each label in here is only referenced once.
       * You may not reference the same label in two different alias.
       */
      const hydratedLabel = hydrateLabel(label)
      /* prettier-ignore */
      const labelAlias: string[] = withDefault([], config[label].alias)
        .filter((aliasName) => currentLabelsNames.includes(aliasName))
      const existingLabel = currentLabelsNames.includes(label)
      const labelIsAliased = labelAlias.length !== 0

      /**
       * New labels:
       *  - doesn't exist yet
       *  - isn't aliased
       */
      if (!existingLabel && !labelIsAliased) {
        added.push(hydratedLabel)
        continue
      }

      /**
       * Updated labels:
       *  - either color or description has changed.
       */

      const labelHasChanged = currentLabels.some(
        (cLabel) =>
          // Must have the same name.
          cLabel.name === hydratedLabel.name &&
          // Description or color might have changed.
          (cLabel.description !== hydratedLabel.description ||
            cLabel.color !== hydratedLabel.color),
      )

      if (existingLabel && labelHasChanged && !labelIsAliased) {
        changed.push(hydratedLabel)
        continue
      }

      /**
       * Aliased labels:
       *  - add an existing label to changed labels
       *  - this one shouldn't be removed afterwards since we
       *    are renaming it.
       */
      const existingAliasedLabel = labelAlias.find((alias) =>
        currentLabelsNames.includes(alias),
      )

      for (const alias of labelAlias) {
        /**
         * We should rename the label to an existing one if the new
         * one doesn't exist yet.
         */
        if (alias === existingAliasedLabel && !existingLabel) {
          changed.push({
            ...hydratedLabel,
            old_name: alias,
          })
        } else {
          aliased.push({
            ...hydratedLabel,
            old_name: alias,
          })
        }
      }

      /* End of label investigation */
    }

    for (const label of currentLabels) {
      /* Indicates that a label has been removed */
      const labelRemoved = !config.hasOwnProperty(label.name)
      const labelRenamed = changed.find(
        ({ old_name }) => old_name === label.name,
      )

      /**
       * Remove all the labels that weren't renamed (changed)
       * as part of the aliasing.
       */
      if (labelRemoved && !labelRenamed) {
        removed.push({
          name: label.name,
          color: label.color,
        })
        continue
      }
    }

    return {
      added,
      changed,
      removed,
      aliased,
    }
  }

  /* Helper functions */

  function hydrateLabel(name: string): GithubLabel {
    return {
      name,
      color: config[name]!.color,
      description: config[name]!.description,
    }
  }
}
