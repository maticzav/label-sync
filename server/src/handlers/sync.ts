/**
 * This file contains the flow for handling label syncing.
 */

import { ProbotOctokit } from 'probot'

import * as config from '../config'
import { Maybe } from '../data/maybe'
import * as gh from '../github'
import { withDefault } from '../utils'

// MARK: - Types

type Octokit = InstanceType<typeof ProbotOctokit>

// MARK: - Methods

type HandleLabelSyncParams = {
  /**
   * Owner of the repository.
   */
  owner: string
  /**
   * Repository name.
   */
  repo: string
  /**
   * LabelSync configuration that we should sync.
   */
  config: config.LSCRepository
}

export type SyncResult =
  | { success: true; diff: Changes }
  | { success: false; message: string }

/**
 * Performs a labels sync with configuration.
 */
export async function sync(
  github: Octokit,
  params: HandleLabelSyncParams,
): Promise<SyncResult> {
  const difference = await changes(github, {
    owner: params.owner,
    repo: params.repo,
    labels: params.config.labels,
  })

  if (difference === null)
    return { success: false, message: `Could not calculate the change.` }

  /**
   * 1. Add new labels.
   * 2. Perform updates.
   * 3. Alias labels.
   * 4. Remove labels.
   */
  try {
    await Promise.all([
      gh.addLabelsToRepository(github, {
        repo: params.repo,
        owner: params.owner,
        labels: difference.additions,
        persist: true,
      }),
      gh.updateLabelsInRepository(github, {
        repo: params.repo,
        owner: params.owner,
        labels: difference.updates,
        persist: true,
      }),
      gh.aliasLabelsInRepository(github, {
        repo: params.repo,
        owner: params.owner,
        labels: difference.aliases,
        persist: true,
      }),
      gh.removeLabelsFromRepository(github, {
        repo: params.repo,
        owner: params.owner,
        labels: difference.removals,
        persist: params.config?.config?.removeUnconfiguredLabels || false,
      }),
    ])

    return { success: true, diff: difference }
  } catch (err) /* istanbul ignore next */ {
    return { success: false, message: err.message }
  }
}

/**
 * Respresents the changes between two configurations.
 */
export type Changes = {
  additions: gh.GithubLabel[]
  updates: gh.GithubLabel[]
  removals: gh.GithubLabel[]
  aliases: gh.GithubLabel[]
}

type ChangesParams = {
  repo: string
  owner: string
  labels: config.LSCRepository['labels']
}

/**
 * Finds the changes between configuration and the current state of the
 * labels on GitHub.
 */
export async function changes(
  github: Octokit,
  params: ChangesParams,
): Promise<Maybe<Changes>> {
  const remoteLabels = await gh.getRepositoryLabels(github, {
    repo: params.repo,
    owner: params.owner,
  })

  /* istanbul ignore if */
  if (remoteLabels === null) return null

  return diff(params.labels, remoteLabels)
}

/**
 * Calculates the diff of two label configurations.
 */
function diff(
  config: config.LSCRepository['labels'],
  current: gh.GithubLabel[],
): Changes {
  /**
   * Label may fall into four classes:
   *  - it may be new: not on Github and no other label
   *    in the configuration references it
   *  - it may be updated: name remains the same but color
   *    or description have changed
   *  - it may be aliased: one of the labels references a
   *    label that exists on Github but is not present in
   *    the configuration anymore
   *  - it may be removed: no labels alias it and it is
   *    not in the configuration
   *
   * In light of that we diff labels so that:
   *  - added: contains all the labels we should add
   *  - changed: contains all the labels that we should rename,
   *      change description or change color
   *  - aliased: contains all the labels that we should cleverly
   *      reference in the issues
   *  - removed: contains all the labels that we should remove
   */

  const currentLabelsMap = new Map<string, gh.GithubLabel>()

  for (const label of current) {
    currentLabelsMap.set(label.name, label)
  }

  /* Labels */

  let added: gh.GithubLabel[] = []
  let changed: gh.GithubLabel[] = []
  let aliased: gh.GithubLabel[] = []
  let removed: gh.GithubLabel[] = []

  /* Calculate differences */

  // TODO: you may not split label alias

  for (const label of config.keys()) {
    /**
     * We know that each label inhere is only referenced once.
     * You may not reference the same label in two different alias.
     */
    const hydratedLabel = hydrateLabel(label)

    const existingLabel = hydratedLabel.old_name !== undefined
    /* prettier-ignore */
    const labelAlias: string[] = withDefault([], config.get(label)?.alias)
        .filter((aliasName) => currentLabelsMap.has(aliasName))
    const labelIsAliased = labelAlias.length !== 0

    /**
     * New labels:
     *  - doesn't exist yet
     *  - isn't aliased
     */
    if (!existingLabel && !labelIsAliased) {
      added.push(hydratedLabel)
      continue
    }

    /**
     * Updated labels:
     *  - either color or description has changed.
     */

    const labelHasChanged = current.some((cLabel) => {
      // Must have the same name.
      const sameName = cLabel.name === hydratedLabel.name

      /**
       * Description is optional. It has changed if:
       *  - it wasn't defined and now is
       *  - it was defined and has changed
       */
      const descriptionChanged =
        withDefault('', cLabel.description) !==
        withDefault('', hydratedLabel.description)

      /**
       * Color of a label is always defined. We check whether it has changed.
       */
      const colorChanged = cLabel.color !== hydratedLabel.color

      return sameName && (descriptionChanged || colorChanged)
    })

    if (existingLabel && labelHasChanged && !labelIsAliased) {
      changed.push(hydratedLabel)
      continue
    }

    /**
     * Aliased labels:
     *  - add an existing label to changed labels
     *  - this one shouldn't be removed afterwards since we
     *    are renaming it.
     */
    const existingAliasedLabel = labelAlias.find((alias) =>
      currentLabelsMap.has(alias),
    )

    for (const alias of labelAlias) {
      /**
       * We should rename the label to an existing one if the new
       * one doesn't exist yet.
       */
      if (alias === existingAliasedLabel && !existingLabel) {
        changed.push({
          ...hydratedLabel,
          old_name: alias,
        })
      } else {
        aliased.push({
          ...hydratedLabel,
          old_name: alias,
        })
      }
    }

    /* End of label investigation */
  }

  for (const label of current) {
    /* Indicates that a label has been removed */
    const labelRemoved = !config.has(label.name)
    const labelRenamed = changed.find(({ old_name }) => old_name === label.name)

    /**
     * Remove all the labels that weren't renamed (changed)
     * as part of the aliasing.
     */
    if (labelRemoved && !labelRenamed) {
      removed.push({
        name: label.name,
        color: label.color,
      })
      continue
    }
  }

  return {
    additions: added,
    updates: changed,
    removals: removed,
    aliases: aliased,
  }

  /* Helper functions */

  function hydrateLabel(name: string): gh.GithubLabel {
    const oldLabel = currentLabelsMap.get(name)

    return {
      /* Naming */
      old_name: oldLabel?.name,
      name,
      /* Description */
      old_description: oldLabel?.description,
      description: config.get(name)!.description,
      /* Color */
      old_color: oldLabel?.color,
      color: config.get(name)!.color,
    }
  }

  //
}

/**
 * Tells whether report has changed.
 */
export function changed(diff: Changes): boolean {
  const changes =
    diff.additions.length +
    diff.aliases.length +
    diff.updates.length +
    diff.removals.length

  return changes > 0
}
