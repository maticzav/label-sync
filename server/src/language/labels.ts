import ml from 'multilines'
import os from 'os'
import { stringifyUrl } from 'query-string'

import { GithubLabel } from '../github'
import { LabelSyncReport } from '../handlers/labels'
import { withDefault, not } from '../utils'
import { isNull } from 'util'

/**
 * Concatenates multiple reports into a human readable string.
 *
 * @param reports
 */
export function generateHumanReadableReport(
  reports: LabelSyncReport[],
): string {
  const humanReadableReports = reports
    .filter(reportHasChanges)
    .sort(orderReports)
    .map(parseLabelSyncReport)
  const report = joinReports(humanReadableReports)

  const unchangedReports = reports.filter(not(reportHasChanges))
  const unchangedReposReport = ulOfUnchangedReports(unchangedReports)

  return ml`
  | ## LabelSync Overview 
  |
  | ### Changes
  | 
  | ${report}
  |
  | ---
  |
  | ### Unchanged repositories
  |
  | ${unchangedReposReport}
  `
}

/**
 * Label report ordering function.
 */
function orderReports(a: LabelSyncReport, b: LabelSyncReport): number {
  /* Show Successes first. */
  if (a.status === 'Success') return -1
  return 0
}

/**
 * Tells whether report has changed.
 */
function reportHasChanges(report: LabelSyncReport): boolean {
  switch (report.status) {
    case 'Success': {
      return (
        report.additions.length +
          report.aliases.length +
          report.removals.length +
          report.updates.length >
        0
      )
    }
    case 'Failure': {
      /* Failure should always be treated as report with changes. */
      return true
    }
  }
}

function parseLabelSyncReport(report: LabelSyncReport): string {
  switch (report.status) {
    case 'Success': {
      const { removeUnconfiguredLabels } = report.config.config

      /* Sections */

      const sections: (string | null)[] = [
        /* Repository name */
        `#### ${parseRepoName(report.repo)}`,
        /* New labels */
        ulOfLabels(
          report.additions,
          singular(
            ':sparkles: 1 new label.',
            (n) => `:sparkles: ${n} new labels.`,
          ),
        ),
        /* Updated labels */
        ulOfLabels(
          report.updates,
          singular(
            ':lipstick: 1 label changed.',
            (n) => `:lipstick: ${n} changed labels.`,
          ),
        ),
        /* Aliased labels */
        ulOfLabels(
          report.aliases,
          singular(
            ':performing_arts: 1 new label alias.',
            (n) => `:performing_arts: ${n} aliases.`,
          ),
        ),
        /* Unconfigured labels and removals */
        (() => {
          if (removeUnconfiguredLabels) {
            return ulOfLabels(
              report.removals,
              singular(
                ':coffin: 1 removed label.',
                (n) => `:coffin: ${n} removed labels.`,
              ),
            )
          }
          return ulOfLabels(
            report.removals,
            singular(
              ':mailbox_with_mail: 1 label unconfigured.',
              (n) => `:mailbox_with_mail: ${n} unconfigured labels.`,
            ),
          )
        })(),
        /* Message about strictness */
        (() => {
          if (!removeUnconfiguredLabels && report.removals.length === 0) {
            return '> You have no unconfigured labels - you could make this repository `strict`.'
          }
          return null
        })(),
      ]

      return sections.filter(not(isNull)).join(os.EOL)
    }
    case 'Failure': {
      return ml`
      | #### ${parseRepoName(report.repo)}
      |
      | ${report.message}
      `
    }
  }
}

/**
 * Let's you create singular and plural versions of a string.
 */
function singular(
  singular: string,
  plural: (val: number) => string,
): (val: number) => string {
  return (n: number) => {
    if (n == 1) return singular
    return plural(n)
  }
}

/**
 * Parses the repo name with backticks.
 *
 * @param name
 */
function parseRepoName(name: string): string {
  return ['`', name, '`'].join('')
}

/**
 * Creates a human readable list of labels.
 */
function ulOfLabels(
  labels: GithubLabel[],
  summary: (n: number) => string,
): string | null {
  if (labels.length == 0) return null
  return ml`
  | <details>
  |   <summary>${summary(labels.length)}</summary>
  |
  | ${labels.map(label).join(os.EOL)}
  |
  | </details>
  |
  `
}

/**
 * Creates a label out of abstract object.
 */
function label(label: GithubLabel): string {
  if (label.old_name) {
    /* prettier-ignore */
    return ` * ${badge({ name: label.old_name, color: "inactive" })} â†’ ${badge(label)}`
  }
  return ` * ${badge(label)}`
}

/**
 * Creates a list of unchanged reports.
 */
function ulOfUnchangedReports(reports: LabelSyncReport[]): string {
  if (reports.length === 0) return `No unchanged repositories.`

  const repos = reports
    .map((report) => ` * ${parseRepoName(report.repo)}`)
    .join(os.EOL)

  return ml`
  | <details>
  |   <summary>${reports.length} repositories.</summary>
  |
  | ${repos}
  |
  | </details>
  `
}

/**
 * Creates a colorful badge for the label.
 */
function badge(props: { name: string; color: string }): string {
  const url = stringifyUrl({
    url: 'https://img.shields.io/static/v1',
    query: {
      label: '',
      message: props.name,
      color: props.color,
    },
  })

  return `![${props.name}](${url} "${props.name}")`
}

/**
 * Joins reports with a separator.
 * @param reports
 */
function joinReports(reports: string[]): string {
  const separator = ['\n', '---', '\n'].join(os.EOL)
  return reports.join(separator)
}
