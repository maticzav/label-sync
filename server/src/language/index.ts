import ml from 'multilines'
import os from 'os'
import prettier from 'prettier'

import { Maybe } from '../data/maybe'
import * as sync from '../handlers/sync'
import * as configs from '../config'

import { divide, label, list, noun, repository } from './components'

// MARK: - Reporters

export type GenerateReportParams = {
  /**
   * Name of the target repository.
   */
  repo: string
  /**
   * Changes in this configuration.
   */
  changes: Maybe<sync.Changes>
  /**
   * Configuration of the repository.
   */
  config: configs.LSCRepositoryConfiguration
}

/**
 * Generates a human readable report out of PR changes for a list of repositories.
 */
export function generateReport(params: GenerateReportParams[]): string {
  const sorted = params.sort((a, b) =>
    a.changes === null && b.changes !== null ? -1 : 0,
  )

  // Overview
  /**
   * Change reports of the changed repositories.
   */
  let changed: string[] = []
  /**
   * Names of the unchanged repositories.
   */
  let unchanged: string[] = []
  /**
   * Names of the repositories that errored.
   */
  let errored: string[] = []

  for (const report of sorted) {
    if (report.changes === null) {
      errored.push(report.repo)
      continue
    }

    // Unchanged
    if (!sync.changed(report.changes)) {
      unchanged.push(report.repo)
      continue
    }

    // Changed repositories
    changed.push(
      generateChangesReport({
        repo: report.repo,
        changes: report.changes,
        config: report.config,
      }),
    )
  }

  // Compose a structured report.
  let sections: string[] = [
    `## :label: Here's what's going to change when you merge`,
    ...changed,
  ]

  if (unchanged.length > 0) {
    const unchanges = list({
      items: unchanged,
      summary: noun(
        `one unchanged repository`,
        (n) => `${n} unchanged repositories`,
      ),
    })

    sections.push(ml`
    | ### You haven't made any changes in these repositories
    |
    | ${unchanges}
    `)
  }

  if (errored.length > 0) {
    const errors = list({
      items: errored,
      summary: noun(
        `couldn't load one repository`,
        (n) => `couldn't load ${n} repositories`,
      ),
    })

    sections.push(ml`
    | ### We couldn't fetch labels from these repositories:
    |
    | ${errors}
    `)
  }

  // Final touches.
  const report = divide(sections)
  return prettier.format(report, { parser: 'markdown' })
}

// MARK: - Chunks

type ChangesReportParams = {
  repo: string
  changes: sync.Changes
  config: configs.LSCRepositoryConfiguration
}

/**
 * Outlines the changes in a human-readable format.
 */
export function generateChangesReport(params: ChangesReportParams): string {
  let sections: Maybe<string>[] = []

  // Title
  sections.push(`#### ${repository(params.repo)}`)

  // New labels
  sections.push(
    list({
      items: params.changes.additions.map(label),
      summary: noun(
        ':sparkles: one new label',
        (n) => `:sparkles: ${n} new labels`,
      ),
    }),
  )

  // Changed labels
  sections.push(
    list({
      items: params.changes.updates.map(label),
      summary: noun(
        ':lipstick: one label changed',
        (n) => `:lipstick: ${n} changed labels`,
      ),
    }),
  )

  // Renamed labels
  sections.push(
    list({
      items: params.changes.updates.map(label),
      summary: noun(
        ':performing_arts: one label renamed',
        (n) => `:performing_arts: ${n} labels renamed`,
      ),
    }),
  )

  // Removed labels
  if (params.config.removeUnconfiguredLabels) {
    sections.push(
      list({
        items: params.changes.removals.map(label),
        summary: noun(
          ':coffin: one label removed',
          (n) => `:performing_arts: ${n} labels renamed`,
        ),
      }),
    )
  }

  // Unconfigured labels.
  if (!params.config.removeUnconfiguredLabels && params.changes.removals) {
    sections.push(
      list({
        items: params.changes.removals.map(label),
        summary: noun(
          ':mailbox_with_mail: one unconfigured label',
          (n) => `:mailbox_with_mail: ${n} unconfigured labels`,
        ),
      }),
    )

    sections.push(`> Make repository string to remove unconfigured labels.`)
  }

  // Create a report
  const report = sections.join(os.EOL)

  return prettier.format(report, { parser: 'markdown' })
}

type ErrorReportParams = {
  repository: string
  message: string
}

/**
 * Outlines an error encountered in execution.
 */
export function generateErrorReport(params: ErrorReportParams): string {
  return ml`
  | #### ${repository(params.repository)}
  |
  | > ❗️ LabelSync encountered a problem syncing this repository.
  |
  | ${params.message}
  `
}
