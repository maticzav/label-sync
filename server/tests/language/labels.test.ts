import { LabelSyncReport } from '../../src/handlers/labels'
import { generateHumanReadableReport } from '../../src/language/labels'

describe('language:', () => {
  test('correctly generates human readable report', () => {
    const reports: LabelSyncReport[] = [
      {
        status: 'Failure',
        owner: 'maticzav',
        repo: 'labelsync',
        message: `Couldn't make a diff of labels.`,
        config: {
          labels: {
            'bug/0': {
              color: 'ff',
            },
            'bug/1': {
              color: '00',
            },
            'bug/2': {
              color: '33',
            },
          },
          config: {
            removeUnconfiguredLabels: false,
          },
        },
      },
      {
        status: 'Success',
        owner: 'maticzav',
        repo: 'changed',
        additions: [
          {
            name: 'bug/2',
            color: '33',
            default: false,
          },
        ],
        updates: [
          {
            name: 'bug/1',
            color: '00',
            default: false,
          },
          {
            old_name: 'label/old',
            name: 'label/new',
            color: '00',
            default: false,
          },
        ],
        removals: [
          {
            name: 'bug/3',
            color: 'ff',
            default: false,
          },
        ],
        aliases: [
          {
            old_name: 'label/alias',
            name: 'label/new-alias',
            color: '00',
            default: false,
          },
        ],
        config: {
          labels: {
            'bug/0': {
              color: 'ff',
            },
            'bug/1': {
              color: '00',
            },
            'bug/2': {
              color: '33',
            },
          },
          config: {
            removeUnconfiguredLabels: true,
          },
        },
      },
    ]

    expect(generateHumanReadableReport(reports)).toMatchSnapshot()
  })
})
