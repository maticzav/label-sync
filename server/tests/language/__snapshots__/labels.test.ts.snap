// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`language: correctly generates human readable report 1`] = `
"## LabelSync Overview 

#### \\\\\`labelsync\\\\\`

Couldn't make a diff of labels.
---
### \`changed\`

__New labels:__
 * bug/2

__Changed labels:__
 * bug/1
 * label/old => label/new

__Aliased labels:__
 * label/alias => label/new-alias

__Removed labels:__
 * bug/3"
`;
